### PR TITLE
PointNet example

### PR DESCRIPTION
**Pull request by dgriffiths3**
Adds an implementation of the seminal [PointNet](https://arxiv.org/abs/1612.00593) point cloud network paper. Includes code to download ModelNet10 dataset and process in a jupyter notebook. Works in colab and local jupyter.

### CODE REVIEW

File: examples/vision/pointnet.py [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+import numpy as np
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+from tensorflow.keras import backend as K
+
+"""
+## Load dataset
+"""
+
```
Comment (fchollet) [inline examples/vision/pointnet.py:40]: No need for the separation between these 2 text blocks

File: examples/vision/pointnet.py [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+"""
+We use the ModelNet10 model dataset, the smaller 10 class version of the ModelNet40
+dataset. First download the data:
+"""
+
+"""shell
+!curl -O http://3dvision.princeton.edu/projects/2014/3DShapeNets/ModelNet10.zip
```
Comment (fchollet) [inline examples/vision/pointnet.py:47]: Prefer using `get_file` to download and cache the file, and unzip it in Python. Having shell commands means the script can't run as a standalone Python file, so shell commands should only be used when there's no alternative.

File: examples/vision/pointnet.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+"""
+To generate a `tf.data.Dataset()` we need to first parse through the ModelNet data
+folders. Each mesh is loaded and sampled into a point cloud before being added to a
+standard python list and converted to a `numpy` array. We also store the current
+enumerate index value as the object label.
+"""
+
+
+def parse_dataset(n_pts=2048):
```
Comment (fchollet) [inline examples/vision/pointnet.py:75]: Prefer using fully spelled-out variable names, like `num_points`

File: examples/vision/pointnet.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+
+    train_pts = []
```
Comment (fchollet) [inline examples/vision/pointnet.py:77]: train_points, etc.

File: examples/vision/pointnet.py [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+    train_labels = []
+    test_pts = []
+    test_labels = []
+
+    folders = glob.glob("ModelNet10/[!README]*")
+
+    for i, folder in enumerate(folders):
+
```
Comment (fchollet) [inline examples/vision/pointnet.py:85]: Remove blank line

File: examples/vision/pointnet.py [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+"""
+To generate a `tf.data.Dataset()` we need to first parse through the ModelNet data
+folders. Each mesh is loaded and sampled into a point cloud before being added to a
+standard python list and converted to a `numpy` array. We also store the current
+enumerate index value as the object label.
+"""
+
+
+def parse_dataset(n_pts=2048):
+
```
Comment (fchollet) [inline examples/vision/pointnet.py:76]: Remove blank line

File: examples/vision/pointnet.py [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+"""
+Set the number of points to sample and batch size and parse the dataset. This can take
+~5minutes to complete.
+"""
+
+N_PTS = 2048
+NUM_CLASSES = 10
+BATCH_SIZE = 32
+
+X_train, X_test, y_train, y_test = parse_dataset(N_PTS)
```
Comment (fchollet) [inline examples/vision/pointnet.py:115]: Use lowercase for variable names, e.g. `x_train`

File: examples/vision/pointnet.py [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+        np.array(test_labels),
+    )
+
+
+"""
+Set the number of points to sample and batch size and parse the dataset. This can take
+~5minutes to complete.
+"""
+
+N_PTS = 2048
```
Comment (fchollet) [inline examples/vision/pointnet.py:111]: NUM_POINTS, etc

File: examples/vision/pointnet.py [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+"""
+Our data can now be read into a `tf.data.Dataset()` object. We set the shuffle buffer
+size to the entire size of the dataset as prior to this the data is ordered by class.
+Data augmentation is important when working with point cloud data. We create a
+augmentation function to jitter, shuffle and rotate the train dataset.
+"""
+
+
+def augment(pts, label):
+
```
Comment (fchollet) [inline examples/vision/pointnet.py:126]: Remove blank line

File: examples/vision/pointnet.py [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+    theta = np.random.uniform(-np.pi, np.pi)
+    c = np.cos(theta)
+    s = np.sin(theta)
+    R = tf.constant([[c, -s, 0.0], [s, c, 0.0], [0.0, 0.0, 1.0]], tf.float64)
+    pts = tf.einsum("...ij,...kj->...ki", R, pts)
+
+    return pts, label
+
+
+train_dataset = tf.data.Dataset.from_tensor_slices((X_train, y_train))
```
Comment (fchollet) [inline examples/vision/pointnet.py:141]: Before generating the dataset, it would be useful to have a visualization of what one set of points looks like.

File: examples/vision/pointnet.py [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+test_dataset = tf.data.Dataset.from_tensor_slices((X_test, y_test))
+
+train_dataset = train_dataset.shuffle(len(X_train)).map(augment).batch(BATCH_SIZE)
+test_dataset = test_dataset.shuffle(len(X_test)).batch(BATCH_SIZE)
+
+"""
+### Build a model
+"""
+
```
Comment (fchollet) [inline examples/vision/pointnet.py:150]: Remove blank space

File: examples/vision/pointnet.py [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+"""
+Each convolution and fully-connected layer (with exception for end layers) consits of
+Convolution / Dense -> Batch Normalization -> ReLU Activation. We can make simple
+sub-classed layer to perform all these as a single layer.
```
Comment (fchollet) [inline examples/vision/pointnet.py:154]: You don't need to, you can take a fully functional approach, which is more concise and does not create additional deserialization overhead

File: examples/vision/pointnet.py [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+"""
+
+
+class ConvBN(layers.Layer):
```
Comment (fchollet) [inline examples/vision/pointnet.py:158]: A functional approach is much simpler:

```python
def conv_bn(x, filters):
   x = layers.Conv1D(filters, kernel_size=1, padding="valid")(x)
   x = layers.BatchNormalization(momentum=0.0)(x)
   return layers.Activation('relu')(x)
```

Same for DenseBN

File: examples/vision/pointnet.py [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+    def call(self, inputs):
+
+        x = self.dense(inputs)
+        x = self.bn(x)
+        return self.activation(x)
+
+
+"""
+PointNet consists of two core components. The primary MLP network, and the transformer
+net (t-net). The t-net aims to learn an affine transformation matrix by its own mini
```
Comment (fchollet) [inline examples/vision/pointnet.py:190]: T-Net

File: examples/vision/pointnet.py [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+network. The mini network strongly resembles the main network. The t-net is used twice.
+The first time to transform the input features ($n \times 3$) into a canonical
```
Comment (fchollet) [inline examples/vision/pointnet.py:192]: Do not use LaTeX notation since we don't render it. Instead, just in-line Python code or pseudo code.

File: examples/vision/pointnet.py [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+representation. The second is an affine transformation for alignment in feature space ($n
+\times 64$). As per the original paper we constrain the transformation to be close to an
+orthogonal matrix (i.e. $||X*X^T - I||$ = 0).
+"""
+
+
+class OrthogonalRegularizer(keras.regularizers.Regularizer):
+    def __init__(self, k, l2reg=0.001):
```
Comment (fchollet) [inline examples/vision/pointnet.py:200]: Please add a docstring and use more descriptive argument names

File: examples/vision/pointnet.py [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+        self.k = k
+        self.l2reg = l2reg
+        self.eye = K.eye(self.k)
+
+    def __call__(self, x):
+        x = K.reshape(x, (-1, self.k, self.k))
```
Comment (fchollet) [inline examples/vision/pointnet.py:206]: Prefer using tf.reshape, tf.tensordot, tf.reduce_sum. Don't use backend ops

File: examples/vision/pointnet.py [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+"""
+
+inputs = keras.Input(shape=(N_PTS, 3))
+
+# x = tnet(N_PTS, 3, 'input_tnet')(inputs)
+
+x = ConvBN(64)(inputs)
+x = ConvBN(64)(x)
+
+# x = tnet(N_PTS, 64, 'feat_tnet')(x)
```
Comment (fchollet) [inline examples/vision/pointnet.py:260]: ?

File: examples/vision/pointnet.py [hunk 19]
```diff
@@ -0,0 +1,295 @@
...
+
+"""
+The main network can be then implemented in the same manner where the t-net mini models
+can be dropped in a layers in the graph. Here we replicate the network architecture
+published in the original paper.
+"""
+
+inputs = keras.Input(shape=(N_PTS, 3))
+
+# x = tnet(N_PTS, 3, 'input_tnet')(inputs)
```
Comment (fchollet) [inline examples/vision/pointnet.py:255]: ?

File: examples/vision/pointnet.py [hunk 20]
```diff
@@ -0,0 +1,295 @@
...
+x = ConvBN(64)(x)
+
+# x = tnet(N_PTS, 64, 'feat_tnet')(x)
+
+x = ConvBN(64)(x)
+x = ConvBN(128)(x)
+x = ConvBN(1024)(x)
+
+x = layers.GlobalMaxPooling1D()(x)
+
```
Comment (fchollet) [inline examples/vision/pointnet.py:267]: You can trim a lot of blank space, in general

File: examples/vision/pointnet.py [hunk 21]
```diff
@@ -0,0 +1,295 @@
...
+outputs = layers.Dense(NUM_CLASSES, activation="softmax")(x)
+
+model = keras.Model(inputs=inputs, outputs=outputs, name="pointnet")
+
+model.summary()
+
+"""
+### Train model
+"""
+
```
Comment (fchollet) [inline examples/vision/pointnet.py:283]: Remove blank space

File: examples/vision/pointnet.py [hunk 22]
```diff
@@ -0,0 +1,295 @@
...
+using `.compile()` and `.fit()`.
+"""
+
+model.compile(
+    loss="sparse_categorical_crossentropy",
+    optimizer=keras.optimizers.Adam(learning_rate=0.001),
+    metrics=["sparse_categorical_accuracy"],
+)
+
+model.fit(train_dataset, epochs=10, validation_data=test_dataset)
```
Comment (fchollet) [inline examples/vision/pointnet.py:295]: After training, it would help to visualize some predictions. How long does training take on a K80?

File: examples/vision/pointnet.py (hunk 2 above)
Comment (fchollet) [reply examples/vision/pointnet.py:47]: Actually, nevermind that comment, shell commands should be fine in this case.

File: examples/vision/pointnet.py (hunk 18 above)
Comment (dgriffiths3) [reply examples/vision/pointnet.py:260]: Accident, was just experimenting without T-net.

File: examples/vision/pointnet.py (hunk 2 above)
Comment (dgriffiths3) [reply examples/vision/pointnet.py:47]: Happy to change if you want?

File: examples/vision/pointnet.py (hunk 12 above)
Comment (dgriffiths3) [reply examples/vision/pointnet.py:154]: I was trying to make layer declaration / calling consistent so now it would be for example:
x = conv_bn(x, n)
x = layers.GlobalMaxPooling1D()(x)

Would it also be better to change the T-Net to be fully functional in this style, or better left as their own model objects?

File: examples/vision/pointnet.py (hunk 22 above)
Comment (dgriffiths3) [reply examples/vision/pointnet.py:295]: Validation accuracy usually hit around 80-90% after 10 epochs, which is inline with expected performance. 1 epoch = 30s on google colab gpu (k80?), so 5 minutes.

File: examples/vision/pointnet.py (hunk 12 above)
Comment (fchollet) [reply examples/vision/pointnet.py]: A function that returns a `Model` instance is a perfectly fine pattern, no problem there.

File: examples/vision/pointnet.py [hunk 23]
```diff
@@ -0,0 +1,283 @@
...
+        x = tf.reshape(x, (-1, self.k, self.k))
+        xxt = tf.tensordot(x, x, axes=(2, 2))
+        xxt = tf.reshape(xxt, (-1, self.k, self.k))
+        return tf.reduce_sum(self.l2reg * tf.square(xxt - self.eye))
+
+"""
+ We can define the T-net as a functional model.
+"""
+
+def tnet(num_points, k, name):
```
Comment (fchollet) [inline examples/vision/pointnet.py:186]: Please use a more descriptive name for `k`.

Going back to the comment about using a functional approach -- return a `Model` instance is fine, but in this case it turns out you don't need to do it, because you never use that instance as a model (e.g. `fit()`). So here converting this function to just apply layers to an input (like `conv_bn`) would be beneficial, since it would mean you don't need to pass `num_points`.

Comment (dgriffiths3) [reply examples/vision/pointnet.py:186]: k is the number of features in the final dimension i.e. (batch, num_points, k). Could change to num_feats, or num_channels?

Comment (fchollet) [reply examples/vision/pointnet.py:186]: `num_features` sounds good to me, clear & descriptive.

**Comment by fchollet** [comment]
Also, about the CI check: please run the `black` code formatter on the file.

**Comment by dgriffiths3** [comment]
Swapped this to `keras.utils.get_file()` as otherwise it printed out the unzip for the whole dataset.

**Comment by fchollet** [comment]
> Swapped this to keras.utils.get_file() as otherwise it printed out the unzip for the whole dataset.

Sounds good. In the future if you want to use unzip you can use the quiet option `-qq`.

**Comment by dgriffiths3** [comment]
Did you want any other changes?

**Comment by fchollet** [comment]
Everything looks good -- just merged. Thank you for the great contribution!
